### DISCUSSION TITLE
PetAscii font special characters show up as '?'

### DISCUSSION THREAD

**Discussion by GuyCarver** [ocornut/imgui#5163] [New Users Build/Link/Run issues ONLY!]
I've loaded C64_Pro_Mono-STYLE.ttf and created an ImGui::Text() widget pointing to a char* buffer filled with values from 0-255.
What I see is only the standard alpha/numeric characters with everything else showing up as "?", the fallback character.  I'd like some help figuring out how to modify the system to recognize more characters as legitimate.

**Comment by PathogenDavid** [comment]
You need to explicitly opt in to loading characters outside the basic character set. See [this FAQ entry](https://github.com/ocornut/imgui/blob/f5c5926fb91764c2ec0e995970818d79b5873d42/docs/FAQ.md#q-how-can-i-display-and-input-non-latin-characters-such-as-chinese-japanese-korean-cyrillic) for details.

> created an ImGui::Text() widget pointing to a char* buffer filled with values from 0-255.

Rather than doing this, you can view which glyphs are loaded using the metrics window (via either `ShowMetricsWindow` or Tools > Metrics in the demo window):

![Metrics window screenshot](https://user-images.githubusercontent.com/278957/161395541-51136e78-fbf1-4f4e-843d-9d08a1065695.png)

**Reply by GuyCarver** [reply]
I have looked at the glyphs and it looks like this:
![c64font](https://user-images.githubusercontent.com/2214068/161402113-00f548cf-8f25-40d8-b50b-c3d874df0dbb.jpg)

I added this code.  I also attempted to call pC64Font->SetGlyphVisible().  The characters are still '?'.
```
ImVector<ImWchar> ranges;
ImFontGlyphRangesBuilder builder;
builder.AddText("Hello world");                        // Add a string (here "Hello world" contains 7 unique characters)
builder.AddChar(0x00a3);
builder.AddChar(0x00b2);
builder.AddChar(0x00b3);
builder.AddChar(0x00b4);
builder.AddChar(0x00bf);
builder.AddChar(0x00c0);
builder.AddChar(0x00c1);
builder.AddChar(0x00c2);
builder.AddChar(0x00c3);
builder.AddChar(0x00c4);
builder.AddChar(0x00c5);
builder.AddRanges(io.Fonts->GetGlyphRangesDefault()); // Add one of the default ranges
builder.BuildRanges(&ranges);                          // Build the final result (ordered ranges with all the unique characters submitted)
pC64Font = io.Fonts->AddFontFromFileTTF("C64_Pro_Mono-STYLE.ttf", FONTSIZE, nullptr, ranges.Data);
```

**Reply by PathogenDavid** [reply]
> I have looked at the glyphs and it looks like this:

From what I see, all of the characters you're adding to your ranges are visible in that screenshot. Is there something missing?

> I also attempted to call pC64Font->SetGlyphVisible().

[You should not call this method.](https://github.com/ocornut/imgui/blob/f5c5926fb91764c2ec0e995970818d79b5873d42/imgui.h#L2881) If your glyphs aren't marked as visible by default there's something wrong with your font. (You can hover them in the metrics window to check, but you wouldn't see `?` if it was wrong.)

> The characters are still '?'.

Which characters?

**Comment by GuyCarver** [comment]
I've fixed using a full 0-255 Ascii character range by adding a FullAscii bool to ImFontConfig which defaults to false. I set it to true for the C64 font.
Other than the ImFontConfig struct and constructor, the only other change required was in ImFont::RenderText()

```       // Decode and advance source
        unsigned int c = (unsigned int)*s & 0xff;
        if ((c < 0x80) || (ConfigData->FullAscii))
        {
            s += 1;
        }
        else
        {
            s += ImTextCharFromUtf8(&c, s, text_end);
            if (c == 0) // Malformed UTF-8?
                break;
        }
```
